File: packages/plugin-sdk-py/src/aisoc_plugin_sdk/connector.py
"""Data-source connector plugin base class."""

from __future__ import annotations

from abc import abstractmethod
from typing import Any, AsyncIterator

from pydantic import BaseModel, Field

from .plugin import AiSOCPlugin, PluginContext


class ConnectorConfig(BaseModel):
    """Configuration for a connector plugin instance."""

    connector_id: str = Field(..., description="Unique connector instance identifier")
    enabled: bool = True
    poll_interval_seconds: int = Field(60, ge=10, description="Polling interval in seconds")
    extra: dict[str, Any] = Field(default_factory=dict, description="Connector-specific settings")


class ConnectorPlugin(AiSOCPlugin):
    """Base class for data-source connector plugins.

    Connectors are responsible for ingesting events from external systems
    (SIEMs, ticketing systems, cloud logs, etc.) and normalising them into
    the AiSOC event format.
    """

    @abstractmethod
    async def test_connection(self, ctx: PluginContext) -> bool:
        """Verify connectivity to the upstream data source. Return True if healthy."""

    @abstractmethod
    async def fetch_events(
        self,
        ctx: PluginContext,
        since: str | None = None,
    ) -> AsyncIterator[dict[str, Any]]:
        """Yield normalised events from the upstream source.

        Args:
            ctx: Plugin context with credentials and config.
            since: ISO-8601 timestamp cursor; fetch events after this point.

        Yields:
            Normalised event dicts conforming to the AiSOC OCSF event schema.
        """

File: packages/plugin-sdk-go/aisoc/connector.go
package aisoc

import "context"

// ConnectorConfig holds per-instance configuration for a connector plugin.
type ConnectorConfig struct {
	ConnectorID           string         `json:"connector_id"`
	Enabled               bool           `json:"enabled"`
	PollIntervalSeconds   int            `json:"poll_interval_seconds"`
	Extra                 map[string]any `json:"extra,omitempty"`
}

// Connector is implemented by plugins that ingest events from external data sources.
type Connector interface {
	Plugin
	// TestConnection verifies that the upstream data source is reachable.
	TestConnection(ctx context.Context, pctx PluginContext) (bool, error)
	// FetchEvents returns a channel of normalised events from the upstream source.
	// The caller is responsible for draining and closing the channel.
	// since is an ISO-8601 timestamp cursor; only events after this point should be returned.
	FetchEvents(ctx context.Context, pctx PluginContext, since string) (<-chan map[string]any, error)
}
